zephyr/esp32s3/src/esp_adc_cal/esp_adc_cal.c

```diff
@@ -0,0 +1,186 @@
+                                                {-1, -1, 1, 0,  0},  //ADC1 atten2
+                                                {-1, -1, 1, -1, 1}   //ADC1 atten3
+                                            };
+const static int32_t adc2_error_sign[4][5] = {
+                                                {-1,  1, 1,  0, 0},  //ADC2 atten0
+                                                {-1, -1, 1,  0, 0},  //ADC2 atten1
+                                                {-1, -1, 1,  0, 0},  //ADC2 atten2
+                                                { 1, -1, 1, -1, 1}   //ADC2 atten3
+                                            };
+
+/* -------------------- Characterization Helper Data Types ------------------ */
+typedef struct {
+    uint32_t voltage;
+    uint32_t digi;
+} adc_calib_data_ver1_t;
+
+typedef struct {
+    char version_num;
+    adc_unit_t adc_num;
+    adc_atten_t atten_level;
+    union {
+        adc_calib_data_ver1_t ver1;
+    } ref_data;
+} adc_calib_info_t;
+
+
+//To get the reference point (Dout, Vin)
+static esp_err_t get_reference_point(int version_num, adc_unit_t adc_num, adc_atten_t atten, adc_calib_info_t *calib_info)
+{
+    assert(version_num == 1);
+    esp_err_t ret;
+
+    calib_info->version_num = version_num;
+    calib_info->adc_num = adc_num;
+    calib_info->atten_level = atten;
+
+    uint32_t voltage = 0;
+    uint32_t digi = 0;
+    ret = esp_efuse_rtc_calib_get_cal_voltage(version_num, ((adc_num == ADC_UNIT_1) ? 0 : 1), atten, &digi, &voltage);
+    assert(ret == ESP_OK);
+    calib_info->ref_data.ver1.voltage = voltage;
+    calib_info->ref_data.ver1.digi = digi;
+    return ret;
+}
+
+esp_err_t esp_adc_cal_check_efuse(esp_adc_cal_value_t source)
+{
+    if (source != ESP_ADC_CAL_VAL_EFUSE_TP_FIT) {
+        return ESP_ERR_NOT_SUPPORTED;
+    }
+    uint8_t adc_encoding_version = esp_efuse_rtc_calib_get_ver();
+    if (adc_encoding_version != 1) {
+        // current version only accepts encoding ver 1.
+        return ESP_ERR_INVALID_VERSION;
+    }
+    return ESP_OK;
+}
+
+/*
+ * Get an expected linear relationship btwn Vin and Dout
+ */
+static void calculate_characterization_coefficients(const adc_calib_info_t *parsed_data, esp_adc_cal_characteristics_t *chars)
+{
+    chars->coeff_a = coeff_a_scaling * parsed_data->ref_data.ver1.voltage / parsed_data->ref_data.ver1.digi;
+    chars->coeff_b = 0;
+    ESP_LOGV(LOG_TAG, "Calib V1, Cal Voltage = %d, Digi out = %d, Coef_a = %d\n", parsed_data->ref_data.ver1.voltage, parsed_data->ref_data.ver1.digi, chars->coeff_a);
+}
+
+esp_adc_cal_value_t esp_adc_cal_characterize(adc_unit_t adc_num,
+        adc_atten_t atten,
+        adc_bits_width_t bit_width,
+        uint32_t default_vref,
+        esp_adc_cal_characteristics_t *chars)
+{
+    (void) default_vref;
+
+    // Check parameters
+    assert((adc_num == ADC_UNIT_1 || adc_num == ADC_UNIT_2, ESP_ADC_CAL_VAL_NOT_SUPPORTED));
```

Comment:
I'm trying to merge your changes with PR 228.  Looks like you uses assert() here instead of returning error codes.  Is that something that we want to keep?